src/analysis/lattice.h

```diff
@@ -42,15 +43,31 @@ constexpr bool is_lattice = has_getBottom<Lattice>&& has_compare<Lattice>&&
 // Represents a powerset lattice which is constructed from a finite set which
 // can be represented by a bitvector. Set elements are represented by
 // FinitePowersetLattice::Element, which represents members present in each
-// element by bits in the bitvector.
-class FinitePowersetLattice {
+// element by bits in the bitvector. Type is the type of a member in a lattice
+// element.
+template<typename Type> class FinitePowersetLattice {
 
   // The size of the set that the powerset lattice was created from. This is
   // equivalent to the size of the Top lattice element.
   size_t setSize;
 
+  // Maps a bitvector index to some element member of Type.
+  // Used to produce initial ordering of element members.
+  std::vector<Type> memberOrdering;
+
+  // Maps an element member of Type to a bitvector index.
+  std::unordered_map<Type, size_t> memberMapping;
+
 public:
-  FinitePowersetLattice(size_t setSize) : setSize(setSize) {}
+  // Takes in an ordered list of all elements belonging to the set to create
+  // the powerset lattice from (i.e. the powerset lattice top element). This
+  // is used for mapping these to bitvector indices.
+  FinitePowersetLattice(std::vector<Type> setMembers)
+    : setSize(setMembers.size()), memberOrdering(setMembers) {
```

Comment:
It probably makes sense to move the vector into the lattice to avoid copying it. I see that you are already doing that in the calling code, but we need this part as well to avoid all copies.

```suggestion
  FinitePowersetLattice(std::vector<Type>&& setMembers)
    : setSize(setMembers.size()), memberOrdering(std::move(setMembers)) {
```